sopel/plugins/jobs.py

```diff
@@ -0,0 +1,91 @@
+# coding=utf-8
+"""Sopel's plugin jobs management.
+
+.. versionadded:: 7.1
+
+.. important::
+
+    This is all fresh and new. Its usage and documentation is for Sopel core
+    development and advanced developers. It is subject to rapid changes
+    between versions without much (or any) warning.
+
+    Do **not** build your plugin based on what is here, you do **not** need to.
+
+"""
+# Copyright 2020, Florian Strzelecki <florian.strzelecki@gmail.com>
+#
+# Licensed under the Eiffel Forum License 2.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import itertools
+import logging
+
+from sopel import tools
+from sopel.tools import jobs
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Scheduler(jobs.JobScheduler):
+    """Scheduler of plugin jobs.
+
+    This scheduler stores plugin jobs and can then run them at regular
+    intervals.
+    """
+    def __init__(self, manager):
+        super(Scheduler, self).__init__(manager)
+        self._jobs = tools.SopelMemoryWithDefault(list)
+
+    def register(self, job):
+        with self._mutex:
+            self._jobs[job.get_plugin_name()].append(job)
+        LOGGER.debug('Job registered: %s', str(job))
+
+    def unregister_plugin(self, plugin_name):
+        """Unregister all the jobs from a plugin.
+
+        :param str plugin_name: the name of the plugin to remove
+        :return: the number of jobs unregistered for this plugin
+        :rtype: int
+
+        All jobs of that plugin will be removed from the scheduler.
+        """
+        unregistered_jobs = 0
+        with self._mutex:
+            jobs_count = len(self._jobs[plugin_name])
+            del self._jobs[plugin_name]
+            unregistered_jobs = unregistered_jobs + jobs_count
+
+        LOGGER.debug(
+            '[%s] Successfully unregistered %d jobs',
+            plugin_name,
+            unregistered_jobs)
+
+        return unregistered_jobs
+
+    def clear_jobs(self):
+        with self._mutex:
+            self._jobs = tools.SopelMemoryWithDefault(list)
+
+        LOGGER.debug('Successfully unregistered all jobs')
+
+    def remove_callable_job(self, callable):
+        """Remove ``callable`` from the job queue."""
```

Comment:
```suggestion
        """Remove ``callable`` from the job queue.

        :param callable: the callable to remove
        :type callable: :term:`callable`
        """
```

It's kinda silly, but consistency is nice. And our new docstring guidelines (#1794) demand param docs.

Reply:
Sadly, `callable` is not a term of the Python glossary. I'll use `function` instead.

Reply:
(docstring updated on the base class)